src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+    {
+        $str = clone $this;
+
+        foreach ($strings as $k => $s) {
+            if ('' !== $s && !preg_match('//u', $s)) {
+                throw new InvalidArgumentException(sprintf('Invalid UTF-8 string at index "%s".', $k));
+            }
+        }
+
+        $str->string = implode($str->string, $strings);
+
+        return $str;
+    }
+
+    abstract public function length(): int;
+
+    public function match(string $pattern, int $flags = 0, int $offset = 0): ?array
+    {
+        if ('g' === $pattern[-1]) {
+            $pattern = substr($pattern, 0, -1);
+            $match = 'preg_match_all';
+        } else {
+            $match = 'preg_match';
+        }
+
+        if ($this->ignoreCase) {
+            $pattern .= 'i';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (false === $ret = $match($pattern.'u', $this->string, $matches, $flags, $offset)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        return $ret ? $matches : null;
+    }
+
+    abstract public function indexOf(string $needle, int $offset = 0): ?int;
+
+    abstract public function indexOfLast(string $needle, int $offset = 0): ?int;
+
+    /**
+     * @return static
+     */
+    public function padBoth(int $length, string $padStr = ' '): self
+    {
+        if ('' === $padStr || !preg_match('//u', $padStr)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $pad = clone $this;
+        $pad->string = $padStr;
+
+        return $this->pad($length, $pad, STR_PAD_BOTH);
+    }
+
+    /**
+     * @return static
+     */
+    public function padLeft(int $length, string $padStr = ' '): self
+    {
+        if ('' === $padStr || !preg_match('//u', $padStr)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $pad = clone $this;
+        $pad->string = $padStr;
+
+        return $this->pad($length, $pad, STR_PAD_LEFT);
+    }
+
+    /**
+     * @return static
+     */
+    public function padRight(int $length, string $padStr = ' '): self
+    {
+        if ('' === $padStr || !preg_match('//u', $padStr)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $pad = clone $this;
+        $pad->string = $padStr;
+
+        return $this->pad($length, $pad, STR_PAD_RIGHT);
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function prepend(string $prefix): self;
```

Comment:
```suggestion
    abstract public function prepend(string ...$prefixes): self;
```

`$str->prepend($a, $b, $c)` vs `$str->prepend($a)->prepend($b)->prepend($c);`